architecture/2025-06-non-k8s-controlplane/2025-06-graph-db-replace-planes.md

```diff
@@ -0,0 +1,370 @@
+* **Access Control:** Access to Kùzu is through the embedded Go driver within the Radius process. Standard Radius authentication and authorization mechanisms will protect the Radius APIs that indirectly interact with Kùzu. There is no direct network exposure of Kùzu in the embedded model.
+* **Input Sanitization:** If any user-provided data is used to construct Cypher queries (even if parameterized), ensure proper parameterization is always used by the DAL to prevent injection vulnerabilities (Kùzu's Go driver should support parameterized queries).
+* **Threat Model:** The Radius threat model must be updated to have a section for the DAL.
+
+### Compatibility (optional)
+
+* **Backward Compatibility:**
+    * For existing Radius deployments using etcd as the graph store, a migration path will be necessary. 
+    * The public Radius API and CLI should remain backward compatible.
+* **Data Format:** The structure of the application graph (apps, resources, properties) should remain conceptually the same, even though the storage backend changes.
+
+### Monitoring and Logging
+
+* **Logging:**
+    * The Graph Abstraction Layer should log all significant operations (e.g., Kùzu queries, errors, transaction boundaries) at appropriate log levels.
+* **Metrics:**
+    * Expose metrics from the DAL in Radius OpenTelemetry:
+        * Number of Kùzu queries (per type: read/write).
+        * Latency of Kùzu queries.
+        * Error rates for Kùzu operations.
+        * Kùzu database size.
+        * Transaction commit/rollback counts.
+
+### Development plan
+
+0.  **Phase 0: DAL (Milestone 0)**
+    * Create the DAL.
+    * Implement CRUD endpoints representing Radius abstraction level graph operations.
+    * Develop initial unit & integration tests for the DAL.
+    * Ensure via debug logging that no components are communicating directly with etcd other than the DAL.
+1.  **Phase 1: Core Integration (Milestone 1)**
+    * Research Kùzu Go driver capabilities and limitations in detail.
+    * Set up Kùzu as an embedded dependency - but as a pluggable architecture where the embedded calls could be swapped for network calls to any Cypher compatible graph database.
+    * Define and implement Kùzu schema creation.
+    * Implement robust error handling and transaction management in the DAL.
+    * Add Kùzu specific tests (schema creation, backup/restore, etc)
+    * Modify Radius init and upgrade processes to trigger appropriate behavior in the DAL.
+4.  **Phase 2: Tooling & Testing (Milestone 2)**
+    * Implement backup/restore CLI commands.
+    * Conduct comprehensive E2E testing, performance testing, and security review of DAL threat model.
+    * Develop documentation for operators and developers.
+3.  **Phase 2: Query Enhancement (Milestone 3 - optional)**
+    * Enhance DAL with more advanced query capabilities (pathfinding, complex traversals, to support new User Stories defined by product).
+
+### Open Questions
```

Comment:
We should consider failure scenarios in which etct/postres is out of sync with the kizu graph.

Reply:
Here's one: if the graph DB is corrupted or lost on a production cluster, and a restore of it loses transactions, the graph will be out of sync with the RP database. Maybe there is no recovery from this type of scenario, but if we are going to have completely separate DBs that have related data, we should understand the impact of what happens when they are out of sync.